### PR TITLE
libserialize: encoding non string key map to json silently produces invalid json instead of failing

### PR DESCRIPTION
**Pull request by oli-obk**
importing object type string key maps is still supported
writing them should be explicit, and can be done as follows

``` rust
let some_tree_map : TreeMap<String, Json> = ...;
Json::Object(some_tree_map).to_writer(&mut writer);
```

related to #8335, #9028, #9142

### CODE REVIEW

**Comment by oli-obk** [comment]
rebasing required quite some changes compared to before. But i think the diffs are not quite as extensive as they were before.

**Comment by oli-obk** [comment]
i'm recompiling rustc from today right now. after that i'll attempt to get this into [rustc-serialize](https://github.com/rust-lang/rustc-serialize)

**Comment by oli-obk** [comment]
might close #8883 but this only does runtime-checks, no compile time checks, the issue is "Compiler accepts deriving(Encodable, Decodable) for TreeMaps with non-str keys"

File: src/libserialize/json.rs [hunk 1]
```diff
@@ -275,6 +275,13 @@ pub enum DecoderError {
     ApplicationError(string::String)
 }
 
+#[derive(Show)]
+pub enum EncoderError {
+    FmtError(fmt::Error),
+    IoError(io::IoError),
```
Comment (alexcrichton) [inline src/libserialize/json.rs]: I think you can remove this variant as it shouldn't be generated any more.

**Comment by alexcrichton** [comment]
Thanks! Just one small nit and otherwise r=me

**Comment by oli-obk** [comment]
r=alexcrichton

**Comment by oli-obk** [comment]
phew, good thing it took me so long to get this ready... it doesn't work!
if a single sub-element of the map key is a valid map key, the flag is set and the map key is accepted.
what i could do is blacklist all invalid elements, but any changes in the future to the Encoder or Decoder trait might mess that up again.

**Comment by oli-obk** [comment]
@alexcrichton it is fixed and tested now, i actually messed up the unit test before, that's why the mistake wasn't caught by it.
make check-stage1 ran through, make check is running right now

**Comment by oli-obk** [comment]
should i squash the fix?

**Comment by alexcrichton** [comment]
Sure!

**Comment by oli-obk** [comment]
done

**Comment by oli-obk** [comment]
@alexcrichton those failures don't seem to be mine, should i rebase or sth?

**Comment by oli-obk** [comment]
rebased

**Comment by oli-obk** [comment]
the only fix required was changing a `0..n` to a `..n`

**Pull request merged**